File: CSD2a/multi_sample_sequencer/src/playhead.py
# Written by Wouter Ensink


# This class resembles the current position in the playback
# Also responsible for the looping
class PlayHead:
    def __init__(self):
        self.position_in_ticks = 0
        self.range_start = 0
        self.range_end = -1

    def rewind(self) -> None:
        self.position_in_ticks = 0

    def set_looping(self, range_start: int, range_end: int) -> None:
        self.range_start = range_start
        self.range_end = range_end

    def stop_looping(self) -> None:
        self.range_end = -1

    def is_looping(self) -> bool:
        return self.range_end > 0

    def advance_tick(self) -> None:
        self.position_in_ticks += 1

        if self.is_looping():
            if self.position_in_ticks >= self.range_end:
                self.position_in_ticks = self.range_start

File: CSD2b/console_synth/source/sequencer/play_head.cpp

// Written by Wouter Ensink

#include <console_synth/sequencer/play_head.h>


void PlayHead::setTickTimeMs (double tickTime)
{
    tickTimeMs = tickTime;
    timeSinceLastTickMs = tickTime;
}

void PlayHead::setDeviceCallbackDurationMs (double duration)
{
    blockDurationMs = duration;
}

void PlayHead::setPositionInTicks (uint64_t position)
{
    currentTick = position;
}

void PlayHead::setTimeSinceLastTickMs (double time)
{
    timeSinceLastTickMs = time;
}


[[nodiscard]] double PlayHead::getTimeSinceLastTickMs() const
{
    return timeSinceLastTickMs;
}


void PlayHead::setLooping (uint64_t start, uint64_t end)
{
    loopingRangeTicks = { start, end };

    if (! (*loopingRangeTicks).contains (currentTick))
    {
        currentTick = start;
    }
}


[[nodiscard]] std::optional<uint64_t> PlayHead::getLoopingStart() const
{
    if (isLooping())
        return loopingRangeTicks->getStart();

    return std::nullopt;
}


[[nodiscard]] std::optional<uint64_t> PlayHead::getLoopingEnd() const
{
    if (isLooping())
        return loopingRangeTicks->getEnd();

    return std::nullopt;
}

[[nodiscard]] uint64_t PlayHead::getCurrentTick() const
{
    return currentTick;
}


[[nodiscard]] double PlayHead::getDeviceCallbackDurationMs() const
{
    return blockDurationMs;
}

[[nodiscard]] double PlayHead::getTickTimeMs() const
{
    return tickTimeMs;
}


[[nodiscard]] bool PlayHead::isLooping() const
{
    return loopingRangeTicks.has_value();
}


void PlayHead::advanceDeviceBuffer()
{
    auto timePoint = tickTimeMs - timeSinceLastTickMs;
    auto tick = currentTick;

    while (timePoint < blockDurationMs)
    {
        timePoint += tickTimeMs;
        tick = getTickAfter (tick);
    }

    timeSinceLastTickMs = blockDurationMs - (timePoint - tickTimeMs);
    // tick now represents the last tick in the current buffer, so it should be one after that
    currentTick = tick;
}


[[nodiscard]] uint64_t PlayHead::getTickAfter (uint64_t tick) const
{
    tick += 1;
    if (isLooping())
        if (tick >= loopingRangeTicks->getEnd())
            tick = loopingRangeTicks->getStart();
    return tick;
}